feat: add macros for DataFusionError variants

Chen-Yuan-Lai opened the pull request:
## Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

- Closes #15491.

## Rationale for this change

<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

## What changes are included in this PR?

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->
As #15491 says, add two macros for 
* External
* ExecutionJoin

Moreover, the macro for `ResourcesExhausted` already existed, so we don't need to add it in the PR

## Are these changes tested?

<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

## Are there any user-facing changes?

<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please add the `api change` label.
-->

comphead commented on datafusion/common/src/error.rs:
```diff
@@ -808,12 +808,18 @@ make_error!(plan_err, plan_datafusion_err, Plan);
 make_error!(internal_err, internal_datafusion_err, Internal);
 
+// Exposes a macro to create `DataFusionError::IoError` with optional backtrace
+make_error!(external_err, external_datafusion_err, External);
```
wasn't there any current usage of this error types it to replace with macros?

Chen-Yuan-Lai replied:
Hi @comphead, I noticed that using `make_error!` isn't suitable for these error variants because `External` and `ExecutionJoin` are designed to wrap original error objects, not just error messages.

Instead, I've implemented custom macros for `External` and  `ExecutionJoin` errors, similar to #15796, and replaced a few use cases as examples. Should we replace all the cases?

comphead reviewed (commented):
The key point of this macros is preserving backtrace which makes debugging easier if you got an error from datafusion among multiple layers of processing. 

The variant below is for non string errors preserving the backtrace

```
#[macro_export]
macro_rules! schema_err {
    ($ERR:expr $(; diagnostic = $DIAG:expr)?) => {{
        let err = $crate::error::DataFusionError::SchemaError(
            $ERR,
            Box::new(Some($crate::error::DataFusionError::get_back_trace())),
        );
        $(
            let err = err.with_diagnostic($DIAG);
        )?
        Err(err)
    }
    };
}
```

Chen-Yuan-Lai commented:
@comphead Thank you for your review. You're right - my current implementation doesn't preserve backtraces.

To preserve backtraces, I think there are two options:
1. **Add Context wrapping**: We could add backtrace by using Context
   ```rust
   let err = DataFusionError::External(Box::new($ERR))
       .context(format!("backtrace: {}", DataFusionError::get_back_trace()));
   ```
    However, this would create double wrapping and might affect error handling logic.

2. **Modify `DataFusionError` enum definitions**: Change error variants to include backtrace fields
    ```rust
    External(GenericError, Option<String>),
    ExecutionJoin(JoinError, Option<String>),
    ```
    But this would need to change the usage of this error type in the codebase

Do you think which approach is better, or would you recommend a different solution?

comphead reviewed (commented):
Thanks @Chen-Yuan-Lai I think second option is good way to go, more over we already have similar invariants 

```
    /// Error returned by arrow.
    ///
    /// 2nd argument is for optional backtrace
    ArrowError(ArrowError, Option<String>),
```

Chen-Yuan-Lai commented:
### Summary of this change

1. **Enhancement of `DataFusionError` Enum Variants**
    * Updated the `ExecutionJoin` and `External` variants to maintain consistent backtrace handling

2. **Addition of Utility Method for Error Processing**
    * Implemented a new method `DataFusionError::box_error_if_needed()` that handles:
       - Automatic boxing for regular error types
       - Prevention of double-boxing for already boxed errors
       - Support for string-based error messages

3. **Improvement of Error-related Macros**
    * Capture and include backtraces
    * Utilize the new `box_error_if_needed()` method
    * Replace all the usage with new macros

comphead commented on datafusion/common/src/error.rs:
```diff
@@ -655,6 +671,20 @@ impl DataFusionError {
+    where
+        E: Into<GenericError>,
+    {
+        err.into()
```
tbh I'm not sure about if this method is needed, can we just inline it into the macros where it is being used?

Chen-Yuan-Lai replied:
Agree, but to prevent the compiler cannot resolve the target type for `e.into()`, I use `GenericError::from(e)` to explicitly provide the target type

comphead reviewed (commented):
thanks @Chen-Yuan-Lai I think we really close to it

alamb commented on datafusion-cli/src/print_options.rs:
```diff
@@ -143,9 +143,7 @@ impl PrintOptions {
-            return Err(DataFusionError::External(
-                "PrintFormat::Table is not implemented".to_string().into(),
-            ));
+            return external_err!("PrintFormat::Table is not implemented");
```
I have noticed the lack of this macro recently too -- thank you for adding it 🙏

alamb commented:
I am a little worried about changing the DataFusionError variant as it which will be a disruptive downstream change (users have to update all match statements that currently match on `DataFusionError` and they won't really see any new functionality

Rather than change the `DataFusionError` variant,  maybe we could model adding a backtrace using `DataFusionError::Context`

So keep `DataFusionError::External` and potentially add a backtrace like
```
DataFusionError::External(..)
  .context(add_backtrace())
```

Or something like that
